design/20220118.certificate-issuance-exponential-backoff.md

```diff
@@ -0,0 +1,144 @@
+# Exponential backoff for certificate issuance
+
+## Summary
+
+Implement a way to apply exponential backoff when retrying failed certificate
+issuances. Issuance in this context refers to the period of time during which
+the `Issuing` condition on a `Certificate` is set to true and for which a new
+set of issuance-specific resources (`CertificateRequest`s, `Order`s, `Challenge`s etc) is
+created.
+
+## Motivation
+
+Currently failed issuances are retried once an hour without a backoff or time limit. This means that 1) continuous failures in large installations can overwhelm external services 2) in case of some external problems (i.e DNS) that cause issuances to fail and last for longer than a couple of hours, rate limits (see [Let'sEncrypt rate limts](https://letsencrypt.org/docs/rate-limits/)) can be easily hit.
+
+## Goals
+
+- Ensure that retrying failed issuances does not overwhelm external services and is less likely to cause issuance being rate-limited by backing off with an exponentially increasing time period. 
+
+- Ensure that when the backoff is being applied, users have a way to find out when the issuance will be next retried and that the backoff mechanism does not introduce extra complexity for debugging and fits in with the already existing 1 hour backoffs
+
+- Ensure that cmctl renew can still be used to manually force an immediate issuance attempt, so that in cases where an issuance was failing due to a setup error (i.e DNS setup) and a user believes that they have fixed it, they have a way to verify the fix without waiting
+
+## Non-goals
+
+- Make the backoff period configurable (extra complexity). For context Kubernetes pod backoff crashloopbackoff period is _not_ configurable (although it is a very demanded feature [k/k#57291](https://github.com/kubernetes/kubernetes/issues/57291))
+
+- Make it possible to reset the backoff period _without_ having a successful issuance first (i.e by manually running `cmctl renew`)
+
+## Proposal
+
+The creation of `CertificateRequest`s for a particular revision of a `Certificate` will happen with a backoff.
+A new `IssuanceAttempts` status field will be added to `Certificate` that will be used to record the number of failed issuances for this revision.
+Similarly to `status.LastFailureTime`, `status.IssuanceAttempts` field will only be set for a currently failing `Certificate` and will be removed after a successful issuance.
+
+`IssuanceAttempts` will be set by `certificates-issuing` controller after a failed issuance by either bumping the already existing value by 1 or setting it to 1 (first failure). In case of a succeeded issuance, `certificates-issuing` controller will ensure that `status.IssuanceAttempts` is not set.
```

Comment:
Would `failedIssuanceAttempts` be clearer? Personally I always get confused with `issuanceAttempts` sort of wording as it doesn't make it clear whether 1 is the 'normal' value (i.e. one attempt is in progress) or if means failure. `failedIssuanceAttempts` seems more precise and I don't _think_ introduces undue complexity?

Reply:
I guess the _precise_ wording would be `failedIssuanceAttemptsForThisRevision` :) There was another discussion about the naming of this field already https://github.com/jetstack/cert-manager/pull/4757#discussion_r786635212
I can see how `failedIssuanceAttempts` makes more sense than `issuanceAttempts` logically, but it is longer and still isn't _exactly_ precise :)
What about `issuancesAttempted`?